README.md
# MMM-GPIO

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/) smart mirror project.

This module allows control of the GPIO pins on the Raspberry Pi in the form of digital outputs, dimmable LED outputs and Button inputs. 
This module also has the ability to trigger pre-configured scenes and animations. 

| Status       | Version | Date       | Maintained? | Minimum MagicMirror² Version |
|:------------ |:------- |:---------- |:----------- |:---------------------------- |
| Experimental | `0.2.0` | 2020-09-24 | No          |`2.2.1`                       |

### Example
![Rainbow colours animation on an RGB ring LED controlled by MMM-GPIO](images/rainbow.gif?raw=true "LED Animation Example")

A button with an RGB Ring LED running a 'rainbow' animation which fades through the colors of the rainbow. 

### Notable Features
1. Control LED's, Relay's, etc, with GPIO connected outputs
1. Trigger actions or module notifications with GPIO connected buttons
1. Button actions can be assigned to 8 different triggers: press, release, double press, double release, triple press, triple release, long press, and long release
1. LED brightness is controlled by Software PWM (Pulse Width Modulation) using DMA (Direct Memory Access) for zero CPU usage while running
1. Create scenes, which are a set of one or more actions, and trigger them with button presses or notifications
1. Create animations by using sets of actions or existing scenes
1. Create triggers that execute a set of actions when the specified module notification is received
1. Execute actions via notifications from other modules. This module responds to the `"GPIO_ACTION"` notification with the payload being the [Action object](#action-configuration-options).
1. Execute actions via GET or POST requests to /MMM-GPIO/requests

### Notable Limitations
1. The Raspberry Pi's internal Pull-Up and Pull-Down resistors are not supported
1. LED dimming cannot be used on GPIO 6 for some Raspberry Pi models (a pi-blaster restriction)

### Dependencies
1. The '[onoff](https://www.npmjs.com/package/onoff)' node module for buttons and digital outputs (Free to install and use - Installed with this module)
1. The '[Pi-Blaster](https://github.com/sarfata/pi-blaster)' program for LED outputs with dimming (Free to install and use - A copy of it comes with this module)

## Table of Contents
- [MMM-GPIO](#mmm-gpio)
  * [Example](#example)
  * [Notable Features](#notable-features)
  * [Notable Limitations](#notable-limitations)
  * [Dependencies](#dependencies)
- [Table of Contents](#table-of-contents)
- [Installation](#installation)
- [Using the module](#Using-the-module)
  * [MagicMirror² Configuration](#MagicMirror-Configuration)
  * [General Configuration Options](#General-Configuration-Options)
  * [Output Configuration Options](#Output-Configuration-Options)
  * [LED Configuration Options](#LED-Configuration-Options)
  * [Button Configuration Options](#Button-Configuration-Options)
  * [Scene Configuration Options](#Scene-Configuration-Options)
  * [Animation Configuration Options](#Animation-Configuration-Options)
  * [Trigger Configuration Options](#Trigger-Configuration-Options)
  * [Action Configuration Options](#Action-Configuration-Options)
  * [Action List](#Action-List)
- [Wiring Buttons and LED's](#Wiring-Buttons-and-LED-s)
- [Updating the Module](#Updating-the-Module)
- [License](#License)


## Installation
To install the module, use the following terminal commands:
1. Navigate to your MagicMirror's modules folder. If you are using the default installation directory, use the command:<br />`cd ~/MagicMirror/modules`
1. Copy the module to your computer:<br />`git clone https://github.com/glitch452/MMM-GPIO.git`
1. Enter the 'MMM-GPIO' directory and Install the node modules:<br />`cd MMM-GPIO && npm install`
1. Make sure the 'Pi-Blaster' program is executable:<br />`chmod +x pi-blaster`

## Using the module

### MagicMirror² Configuration

To use this module, add an 'MMM-GPIO' configuration block to the modules array in the `config/config.js` file.<br />Below is an example of some configuration options.
```js
var config = {
    modules: [
        ...
        {
            module: 'MMM-GPIO',
            config: {
                leds: [
                    { name: "red",   pin: 20, value: 0.233 },
                    { name: "green", pin: 21, value: 0 },
                    { name: "blue",  pin: 22, value: 1 }
                ],
                outputs: [
                    { name: "relay", pin: 18, value: 1 }
                ],
                buttons: [
                    { name: "btn1",  pin: 23,
                        onDoubleRelease: { action: "TOGGLE", name: "red" }
                    },
                    { name: "btn2",  pin: 24, longPressTime: 3000,
                        onRelease: { action: "NOTIFY", notification: "COMMANDS_ACTION", payload: { action: "DISPLAY_TOGGLE" } },
                        onLongPress: { action: "NOTIFY", notification: "COMMANDS_ACTION", payload: { action: "SHUTDOWN" } },
                        longPressAlert: { title: "Shutdown", message: "Hold for 3 seconds to Shut Down the System.", imageFA: "power-off" }
                    },
                ],
                scenes: [
                    { name: "turquoise", actions: [
                        { action: "SET", name: "red", value: 0 },
                        { action: "SET", name: "green", value: 0.65 },
                        { action: "SET", name: "blue", value: 1 }
                    ] }
                ],
                animations: [
                    { name: "slowPulse", repeat: true,
                        preStartActions: [
                            { action: "STOP_ALL" },
                            { action: "SET_SCENE", name: "turquoise", value: 0 }
                        ],
                        onStopActions: { action: "SET_SCENE", name: "turquoise", value: 0 },
                        frames: [
                            { time: 4000, actions: { action: "SET_SCENE", name: "white", value: 1, time: 4000 } },
                            { time: 4000, actions: { action: "SET_SCENE", name: "white", value: 0.05, time: 4000 } }
                        ]
                    }
                ]
                ...
                // See below for more Configuration Options
            }
        },
        ...
    ]
}
```

### General Configuration Options

| Option                  | Details
|------------------------ |--------------
| `outputs`               | *Optional* - A list of output configurations. These outputs can only be set to on or off, there are no in-between values. See the [Output Configuration](#output-configuration-options) section for more details on Output specific configuration options. <br />**Type:** `array`<br />**Default:** `[]`
| `leds`                  | *Optional* - A list of LED output configurations. The LED outputs should be used when PWM dimming is required. See the [LED Configuration](#led-configuration-options) section for more details on LED specific configuration options. <br />**Type:** `array`<br />**Default:** `[]`
| `buttons`               | *Optional* - A list of Button input configurations. See the [Button Configuration](#button-configuration-options) section for more details on Button specific configuration options. <br />**Type:** `array`<br />**Default:** `[]`
| `scenes`                | *Optional* - A list of Scene configurations. See the [Scene Configuration](#scene-configuration-options) section for more details on Scene specific configuration options. <br />**Type:** `array`<br />**Default:** `[]`
| `animations`            | *Optional* - A list of Animation configurations. See the [Animation Configuration](#animation-configuration-options) section for more details on Animation specific configuration options. <br />**Type:** `array`<br />**Default:** `[]`
| `triggers`              | *Optional* - A list of Trigger configurations. See the [Trigger Configuration](#trigger-configuration-options) section for more details on Trigger specific configuration options. <br />**Type:** `array`<br />**Default:** `[]`
| `pinScheme`             | *Optional* - The pin numbering system to use for the input/output `pin` options. See this [interactive pinout diagram](https://pinout.xyz) for more details on pin usage for the Raspberry Pi. <br />**Type:** `string`<br />**Default:** `"BCMv2"`<br />**Options:**<br />- `"BCMv2"` The standard Raspberry Pi GPIO numbering system on current (Rev 2) boards<br />- `"BCMv1"` The standard Raspberry Pi GPIO numbering system on older (Rev 1) boards
| `activeLow`             | *Optional* - A 'global' default setting to use active low mode for the inputs and outputs.  When true, the button actions will be triggered when the GPIO pin is pulled low instead of high, and the output values will be inverted. Note: This option can also be configured for each individual input/output. <br />**Type:** `boolean`<br />**Default:** `false`
| `debounceTimeout`       | *Optional* - A 'global' default setting for the button debounce time, which is the number of milliseconds that the input needs to be stable for the button actions to be triggered. Set to `0` to disable the debounce feature. Note: This option can also be configured for each individual button. <br />**Type:** `number`<br />**Default:** `8`
| `multiPressTimeout`     | *Optional* - A 'global' default setting for the number of milliseconds to wait for another press before triggering a press/release action. Note: This option can also be configured for each individual button. <br />**Type:** `number`<br />**Default:** `325`
| `longPressTime`         | *Optional* - A 'global' default setting for the number of milliseconds to wait before triggering a long press action. Note: This option can also be configured for each individual button. <br />**Type:** `number`<br />**Default:** `4000`
| `clearAlertOnRelease`   | *Optional* - A 'global' default setting to control when a button's long press alert is cleared.  Set to `false` to clear the alert on the long press action, set to `true` to clear the alert on the long release action. Note: This option can also be configured for each individual button. <br />**Type:** `boolean`<br />**Default:** `false`
| `usingPiBlasterService` | *Optional* - If you have setup Pi-Blaster to run as a service on start-up, set this option to `true` to prevent this module from starting its own Pi-Blaster instance. Set to `false` to start Pi-Blaster when this module loads. <br />**Type:** `boolean`<br />**Default:** `false`
| `scriptPath`            | *Optional* - Override the location of the pi-blaster program.  If you already have a pi-blaster installation or want to use a different pi-blaster executable, use this option to set the full path to the file. <br />**Type:** `string`<br />**Default:** `<module_folder>/pi-blaster-pi4-deb10`


### Output Configuration Options
The below options can be used for each Output configuration object in the `config.outputs` array.

| Option                  | Details
|:----------------------- |:-------------
| `name`                  | **REQUIRED** - A name to assign to the output. This name is used as a unique identifier for this output. The name is required to trigger actions for this output. Note: The namespace for Outputs, LED's and Buttons is shared. <br />**Type:** `string`
| `pin`                   | **REQUIRED** - The GPIO Pin number to use for this output. The default pin scheme is the standard Raspberry Pi (BCM) GPIO numbering system for Rev 2 Pi's. See the `pinScheme` option in the [General Configuration](#general-configuration-options) section for more details.<br />**Type:** `number`
| `activeLow`             | *Optional* - Use this option to enable active low mode for the output.  When true, the output values will be inverted. <br />**Type:** `boolean`<br />**Default:** `config.activeLow` (`false`)
| `value`                 | *Optional* - The initial value to assign to the output. The possible values are `0` and `1`, where `0` is 'low' and `1` is 'high'. Note: It is recommended to use pull-up or pull-down resistors in your circuit to guarantee initial state before the module loads. <br />**Type:** `number`<br />**Default:** `0`

### LED Configuration Options
The below options can be used for each LED configuration object in the `config.leds` array.

| Option                  | Details
|:----------------------- |:-------------
| `name`                  | **REQUIRED** - A name to assign to the LED. This name is used as a unique identifier for this LED. The name is required to trigger actions for this LED. Note: The namespace for Outputs, LED's and Buttons is shared. <br />**Type:** `string`
| `pin`                   | **REQUIRED** - The GPIO Pin number to use for this LED. The default pin scheme is the standard Raspberry Pi (BCM) GPIO numbering system for Rev 2 Pi's. See the `pinScheme` option in the [General Configuration](#general-configuration-options) section for more details.<br />**Type:** `number`
| `activeLow`             | *Optional* - Use this option to enable active low mode for the LED.  When true, the LED output values will be inverted. <br />**Type:** `boolean`<br />**Default:** `config.activeLow` (`false`)
| `value`                 | *Optional* - The initial value to assign to the LED output. The possible values are any number between `0` and `1`, where `0` is 'low' and `1` is 'high'. ie. A value of `0.25` would set the LED to 25% brightness. Note: It is recommended to use pull-up or pull-down resistors in your circuit to guarantee initial state before the module loads. <br />**Type:** `number`<br />**Default:** `0`


CHANGELOG.md
# MMM-GPIO Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.2.0] - 2020-09-24

### Changed

- Updated pi-blaster kill command to kill processes before starting a new one
- Updated package-lock

## [0.1.0] - 2020-09-19

Initial Experimental Release of the MMM-GPIO module. 


MMM-GPIO.js
/* global Module */

/**
 * Magic Mirror
 * Module: MMM-GPIO
 *
 * By David Dearden
 * MIT Licensed.
 */

var axis, Log;

/**
 * Register the module with the MagicMirror program
 */
Module.register("MMM-GPIO", {
	
	/**
	 * The default configuration options
	 */
	defaults: {
		outputs: [],
		leds: [],
		buttons: [],
		scenes: [],
		animations: [],
		triggers: [],
		pinScheme: "BCMv2",
		activeLow: false,
		debounceTimeout: 8,
		multiPressTimeout: 325,
		longPressTime: 4000,
		clearAlertOnRelease: false,
		usingPiBlasterService: false,
		scriptPath: null, // Set in self.start() becuase access to self.data.path is needed
		
		developerMode: false
	},
	
	/**
	 * The minimum version of magic mirror that is required for this module to run. 
	 */
	requiresVersion: "2.2.1",
	
	/**
	 * Override the start function.  Set some instance variables and validate the selected 
	 * configuration options before loading the rest of the module.  
	 */
	start: function() {
		var self = this;
		var i;
		self.initialized = false;
		self.instanceID = self.identifier + "_" + Math.random().toString().substring(2);
		self.defaults.scriptPath = self.data.path + "pi-blaster-pi4-deb10";
		self.resources = {};
		self.scenes = {};
		self.animations = {};
		self.triggers = [];
		self.pinList = [];
		self.nameList = [];
		self.validPinSchemes = [ "BCMv1", "BCMv2" ];
		self.pinMapping = [
			{ "BOARD": 10, "BCMv1": 15, "BCMv2": 15, "WPI": 16 },
			{ "BOARD": 11, "BCMv1": 17, "BCMv2": 17, "WPI": 0 },
			{ "BOARD": 12, "BCMv1": 18, "BCMv2": 18, "WPI": 1 },
			{ "BOARD": 13, "BCMv1": 21, "BCMv2": 27, "WPI": 2 },
			{ "BOARD": 15, "BCMv1": 22, "BCMv2": 22, "WPI": 3 },
			{ "BOARD": 16, "BCMv1": 23, "BCMv2": 23, "WPI": 4 },
			{ "BOARD": 18, "BCMv1": 24, "BCMv2": 24, "WPI": 5 },
			{ "BOARD": 19, "BCMv1": 10, "BCMv2": 10, "WPI": 12 },
			{ "BOARD": 21, "BCMv1": 9, "BCMv2": 9, "WPI": 13 },
			{ "BOARD": 22, "BCMv1": 25, "BCMv2": 25, "WPI": 6 },
			{ "BOARD": 23, "BCMv1": 11, "BCMv2": 11, "WPI": 14 },
			{ "BOARD": 24, "BCMv1": 8, "BCMv2": 8, "WPI": 10 },
			{ "BOARD": 26, "BCMv1": 7, "BCMv2": 7, "WPI": 11 },
			{ "BOARD": 29, "BCMv1": null, "BCMv2": 5, "WPI": 21 },
			{ "BOARD": 3, "BCMv1": 0, "BCMv2": 2, "WPI": 8 },
			{ "BOARD": 31, "BCMv1": null, "BCMv2": 6, "WPI": 22 }, // GPIO 6 is banned in PiBlaster becuase it is used for ethernet on some boards
			{ "BOARD": 32, "BCMv1": null, "BCMv2": 12, "WPI": 26 },
			{ "BOARD": 33, "BCMv1": null, "BCMv2": 13, "WPI": 23 },
			{ "BOARD": 35, "BCMv1": null, "BCMv2": 19, "WPI": 24 },
			{ "BOARD": 36, "BCMv1": null, "BCMv2": 16, "WPI": 27 },
			{ "BOARD": 37, "BCMv1": null, "BCMv2": 26, "WPI": 25 },
			{ "BOARD": 38, "BCMv1": null, "BCMv2": 20, "WPI": 28 },
			{ "BOARD": 40, "BCMv1": null, "BCMv2": 21, "WPI": 29 },
			{ "BOARD": 5, "BCMv1": 1, "BCMv2": 3, "WPI": 9 },
			{ "BOARD": 7, "BCMv1": 4, "BCMv2": 4, "WPI": 7 },
			{ "BOARD": 8, "BCMv1": 14, "BCMv2": 14, "WPI": 15 }
		];
		self.triggerActionMapping = {
			onPress: "PRESS",
			onRelease: "RELEASE",
			onDoublePress: "DOUBLE_PRESS",
			onDoubleRelease: "DOUBLE_RELEASE",
			onTriplePress: "TRIPLE_PRESS",
			onTripleRelease: "TRIPLE_RELEASE",
			onLongPress: "LONG_PRESS",
			onLongRelease: "LONG_RELEASE",
		};
		self.validTriggers = Object.keys(self.triggerActionMapping);
		
		if (!axis.isString(self.config.scriptPath) || self.config.scriptPath.length < 1 ) { self.config.scriptPath = self.defaults.scriptPath; }
		if (!axis.isArray(self.config.leds)) { self.config.leds = self.defaults.leds; }
		if (!self.validPinSchemes.includes(self.config.pinScheme)) { self.config.pinScheme = self.defaults.pinScheme; }
		if (!axis.isBoolean(self.config.developerMode)) { self.config.developerMode = self.defaults.developerMode; }
		if (!axis.isBoolean(self.config.activeLow)) { self.config.activeLow = self.defaults.activeLow; }
		if (!axis.isBoolean(self.config.clearAlertOnRelease)) { self.config.clearAlertOnRelease = self.defaults.clearAlertOnRelease; }
		if (!axis.isNumber(self.config.debounceTimeout) || isNaN(self.config.debounceTimeout) || self.config.debounceTimeout < 0 ) { self.config.debounceTimeout = self.defaults.debounceTimeout; }
		if (!axis.isNumber(self.config.multiPressTimeout) || isNaN(self.config.multiPressTimeout) || self.config.multiPressTimeout < 0 ) { self.config.multiPressTimeout = self.defaults.multiPressTimeout; }
		if (!axis.isNumber(self.config.longPressTime) || isNaN(self.config.longPressTime) || self.config.longPressTime < 0 ) { self.config.longPressTime = self.defaults.longPressTime; }
		if (!axis.isBoolean(self.config.usingPiBlasterService)) { self.config.usingPiBlasterService = self.defaults.usingPiBlasterService; }
		
		var resourceTypes = [ "leds", "outputs", "buttons", "scenes", "animations", "triggers" ];
		for (i = 0; i < resourceTypes.length; i++) {
			var typeName = resourceTypes[i];
			if (axis.isObject(self.config[typeName])) { self.config[typeName] = [ self.config[typeName] ]; }
			else if (!axis.isArray(self.config[typeName])) { self.config[typeName] = self.defaults[typeName]; }
		}
		
		// Loop through the provided configurations and add valid ones to the resources
		for (i = 0; i < self.config.leds.length; i++) { self.addResource("LED", self.config.leds[i]); }
		for (i = 0; i < self.config.outputs.length; i++) { self.addResource("OUT", self.config.outputs[i]); }
		for (i = 0; i < self.config.buttons.length; i++) { self.addResource("BTN", self.config.buttons[i]); }
		// Add the scenes
		for (i = 0; i < self.config.scenes.length; i++) { self.addScene(self.config.scenes[i]); }
		// Add the animations
		for (i = 0; i < self.config.animations.length; i++) { self.addAnimation(self.config.animations[i]); }
		// Add the triggers
		for (i = 0; i < self.config.triggers.length; i++) { self.addTrigger(self.config.triggers[i]); }
		
		self.log(("start(): self.data: " + JSON.stringify(self.data)), "dev");
		self.log(("start(): self.config: " + JSON.stringify(self.config)), "dev");
		self.log(("start(): self.resources: " + JSON.stringify(self.resources)), "dev");
		self.log(("start(): self.scenes: " + JSON.stringify(self.scenes)), "dev");
		self.log(("start(): self.animations: " + JSON.stringify(self.animations)), "dev");
		self.log(("start(): self.triggers: " + JSON.stringify(self.triggers)), "dev");
		
		self.sendSocketNotification("INIT", {
			instanceID: self.instanceID,
			usingPiBlasterService: self.config.usingPiBlasterService,
			scriptPath: self.config.scriptPath,
			resources: self.resources,
			scenes: self.scenes,
			animations: self.animations,
			developerMode: self.config.developerMode
		});
		
	},
	
	/**
	 * Validate the resource and add it to the resources list
	 * 
	 * @param type (string) The type of resource
	 * @param resource (object) The resource to validate
	 */
	addResource: function(type, resource) {
		var self = this;
		var i, pin, typeFull, triggerName, actionName;
		
		switch (type) {
			case "LED": typeFull = "LED"; break;
			case "OUT": typeFull = "output"; break;
			case "BTN": typeFull = "button"; break;
			default: return;
		}
		
		if (!axis.isString(resource.name) || resource.name.length < 1) {
			self.log(self.translate("MISSING_RESOURCE_NAME", { pin: resource.pin, type: self.translate(typeFull.toUpperCase()) }), "warn");
			return;
		}
		
		if (self.nameList.includes(resource.name)) {
			self.log(self.translate("DUPLICATE_RESOURCE_NAME", { pin: resource.pin, type: self.translate(typeFull.toUpperCase()), name: resource.name }), "warn");
			return;
		}
		
		pin = resource.pin;
		resource.pin = self.validatePin(resource.pin);
		if (axis.isNull(resource.pin) || (type === "LED" && pin === 6)) {
			self.log(self.translate("INVALID_PIN", { pin: pin, type: self.translate(typeFull.toUpperCase()), name: resource.name }), "warn");
			return;
		}
		
		if (self.pinList.includes(resource.pin)) {
			self.log(self.translate("DUPLICATE_PIN", { pin: resource.pin, type: self.translate(typeFull.toUpperCase()), name: resource.name }), "warn");
			return;
		}
		
		if (!axis.isBoolean(resource.activeLow)) { resource.activeLow = self.config.activeLow; }
		
		var result = { type: type, name: resource.name, pin: resource.pin, activeLow: resource.activeLow };
		
		if (type === "LED") {
			if (!axis.isNumber(resource.value) || isNaN(resource.value)) { resource.value = 0; }
			else if (resource.value < 0) { resource.value = 0; }
			else if (resource.value > 1) { resource.value = 1; }
			
			if (!axis.isNumber(resource.exitValue) || isNaN(resource.exitValue)) { resource.exitValue = null; }
			else if (resource.exitValue < 0) { resource.exitValue = 0; }
			else if (resource.exitValue > 1) { resource.exitValue = 1; }
			
			result.value = resource.value;
			result.exitValue = resource.exitValue;
		} else if (type === "OUT") {
			if (resource.value !== 0 && resource.value !== 1) { resource.value = 0; }
			result.value = resource.value;
		} else if (type === "BTN") {
			if (!axis.isBoolean(resource.clearAlertOnRelease)) { resource.clearAlertOnRelease = self.config.clearAlertOnRelease; }
			if (!axis.isNumber(resource.debounceTimeout) || isNaN(resource.debounceTimeout) || resource.debounceTimeout < 0 ) { resource.debounceTimeout = self.config.debounceTimeout; }
			if (!axis.isNumber(resource.multiPressTimeout) || isNaN(resource.multiPressTimeout) || resource.multiPressTimeout < 0 ) { resource.multiPressTimeout = self.config.multiPressTimeout; }
			if (!axis.isNumber(resource.longPressTime) || isNaN(resource.longPressTime) || resource.longPressTime < 0 ) { resource.longPressTime = self.config.longPressTime; }
			
			if (axis.isObject(resource.longPressAlert) && axis.isString(resource.longPressAlert.message)) {
				result.longPressAlert = { message: resource.longPressAlert.message, title: null, imageFA: null };
				if (axis.isString(resource.longPressAlert.title)) { result.longPressAlert.title = resource.longPressAlert.title; }
				if (axis.isString(resource.longPressAlert.imageFA)) { result.longPressAlert.imageFA = resource.longPressAlert.imageFA; }
			} else {
				result.longPressAlert = null;
			}
			
			for (i = 0; i < self.validTriggers.length; i++) {
				triggerName = self.validTriggers[i];
				actionName = self.triggerActionMapping[triggerName];
				result[actionName] = self.validateActions(resource[triggerName]);
			}
			
			if (result.TRIPLE_PRESS.length > 0 || result.TRIPLE_RELEASE.length > 0) { result.numShortPress = 3; }
			else if (result.DOUBLE_PRESS.length > 0 || result.DOUBLE_RELEASE.length > 0) { result.numShortPress = 2; }
			else if (result.PRESS.length > 0 || result.RELEASE.length > 0) { result.numShortPress = 1; }
			else { result.numShortPress = 0; }
			result.enableLongPress = (result.LONG_PRESS.length > 0 || result.LONG_RELEASE.length > 0);
			
			result.clearAlertOnRelease = resource.clearAlertOnRelease;
			result.debounceTimeout = resource.debounceTimeout;
			result.multiPressTimeout = resource.multiPressTimeout;
			result.longPressTime = resource.longPressTime;
		}
		
		self.pinList.push(resource.pin);
		self.nameList.push(resource.name);
		self.resources[resource.name] = result;
		
	},
	
	/**
	 * Validate the provided scene and add it to the scene list
	 * 
	 * @param scene (object) The scene to validate
	 */
	addScene: function(scene) {
		var self = this;
		
		if (!axis.isString(scene.name) || scene.name.length < 1) {
			self.log(self.translate("MISSING_NAME", { type: self.translate("SCENE") }), "warn");
			return;
		}
		
		var result = { type: "SCN", name: scene.name };
		
		result.actions = self.validateActions(scene.actions);
		
		if (result.actions.length < 1) {
			self.log(self.translate("MISSING_ACTIONS", { type: self.translate("SCENE") }), "warn");
			return;
		}
		
		if (!axis.isNumber(scene.value) || isNaN(scene.value)) { scene.value = 1; }
		else if (scene.value < 0) { scene.value = 0; }
		else if (scene.value > 1) { scene.value = 1; }
		
		result.default = scene.value;
		result.value = 0;
		self.scenes[scene.name] = result;
	},
	
	/**
	 * Validate the provided animation and add it to the animation list
	 * 
	 * @param animation (object) The animation to validate
	 */
	addAnimation: function(animation) {
		var self = this;
		var i, frame, frame_result;
		
		if (!axis.isString(animation.name) || animation.name.length < 1) {
			self.log(self.translate("MISSING_NAME", { type: self.translate("ANIMATION") }), "warn");
			return;
		}
		
		var result = { type: "ANI", name: animation.name, frames: [] };
		
		if (axis.isObject(animation.frames)) { animation.frames = [ animation.frames ]; }
		else if (!axis.isArray(animation.frames)) { animation.frames = []; }
		
		for (i = 0; i < animation.frames.length; i++) {
			frame = animation.frames[i];
			frame_result = { actions: self.validateActions(frame.actions) };
			if (frame_result.actions.length > 0) {
				if (!axis.isNumber(frame.time) || isNaN(frame.time)) { frame.time = 1000; }
				if (frame.time < 0) { frame.time = 0; }
				frame_result.time = frame.time;
				result.frames.push(frame_result);
			}
		}
		
		if (result.frames < 1) {
			self.log(self.translate("MISSING_FRAMES", { type: self.translate("ANIMATION") }), "warn");
			return;
		}
		
		result.repeat = animation.repeat === true ? true : false;
		result.running = false;
		result.preStartActions = self.validateActions(animation.preStartActions);
		result.onStopActions = self.validateActions(animation.onStopActions);
		
		self.animations[animation.name] = result;
	},
	
	/**
	 * Validate the provided trigger and add it to the triggers list
	 * 
	 * @param trigger (object) The trigger to validate
	 */
	addTrigger: function(trigger) {
		var self = this;
		if (!axis.isObject(trigger)) { return; }
		trigger.actions = self.validateActions(trigger.actions);
		if (trigger.actions.length > 0) { self.triggers.push(trigger); }
	},
	
	/**
	 * Validate the provided action(s)
	 * 
	 * @param a (object|array) The action object to validate
	 * @param singleMode (boolean) If true, and 'a' is not an array, a single object is returned, otherwise the object is returned in an array
	 * @return (object|null|array) Returns an array of the given valid action(s).  When in singleMode, returns the action, if valid, otherwise null
	 */
	validateActions: function(a, singleMode) {
		var self = this;
		
		if (axis.isArray(a)) {
			var actions = [];
			for (var i = 0; i < a.length; i++) {
				if (axis.isArray(a[i])) { continue; }
				a[i] = self.validateActions(a[i], true);
				if (!axis.isNull(a[i])) { actions.push(a[i]); }
			}
			return actions;
		}
		
		if (singleMode !== true) { singleMode = false; }
		
		if (!axis.isObject(a) || !axis.isString(a.action) || a.action.length < 1) { return singleMode ? null : []; }
		a.action = a.action.toUpperCase();
		if	(	( a.action !== "STOP_ALL" && a.action !== "NOTIFY" && (!axis.isString(a.name) || a.name.length < 1) ) ||
				( a.action === "NOTIFY" && (!axis.isString(a.notification) || a.notification.length < 1) )
			) { return singleMode ? null : []; }
		
		return singleMode ? a : [ a ];
	},
	
	/**
	 * Validate the provided pin number against the pinMapping list
	 * 
	 * @param pin (number) The pin number to validate
	 * @return (number|null) Returns the pin number, if valid, otherwise null
	 */
	validatePin: function(pin) {
		var self = this;
		var pinObj = self.pinMapping.find(function(val) { return val[this.scheme] === this.pin; }, { scheme: self.config.pinScheme, pin: pin });
		if (axis.isUndefined(pinObj)) {
			return null;
		} else {
			return pinObj[self.config.pinScheme];
		}
	},
	
	/**
	 * Override the socketNotificationReceived function to handle the notifications sent from the node helper
	 * 
	 * @param notification (string) The type of notification sent
	 * @param payload (any) The data sent with the notification
	 */
	socketNotificationReceived: function(notification, payload) {
		var self = this;
		
		// If there is no module ID sent with the notification
		if (!axis.isObject(payload) || !axis.isObject(payload.original) || !axis.isString(payload.original.instanceID)) {
			if (notification === "LOG") {
				if (payload.translate) { self.log(self.translate(payload.message, payload.translateVars), payload.logType); }
				else { self.log(payload.message, payload.logType); }
			} else if (notification === "NOTIFY") {
				self.sendNotification(payload.notification, payload.payload);
			} else if (notification === "INIT") {
				self.initialized = true;
			}
			return;
		}
		
		// Filter out notifications for other instances
		if (payload.original.instanceID !== self.instanceID) {
			self.log(("Notification ignored for ID \"" + payload.original.instanceID + "\"."), "dev");
			return;
		}
		
		if (notification === "LOG") {
			if (payload.translate) { self.log(self.translate(payload.message, payload.translateVars), payload.logType); }
			else { self.log(payload.message, payload.logType); }
		}
	},
	
	/**
	 * Override the notificationReceived function.  
	 * For now, there are no actions based on system or module notifications.  
	 * 
	 * @param notification (string) The type of notification sent
	 * @param payload (any) The data sent with the notification
	 * @param sender (object) The module that the notification originated from
	 */
	notificationReceived: function(notification, payload, sender) {
		var self = this;
		var senderName = axis.isObject(sender) ? sender.name : "undefined";
		self.log(("notificationReceived(): " + notification + " " + JSON.stringify(payload) + " " + senderName), "dev");
		
		if (!self.initialized) { return; }
		
		if (axis.isObject(sender)) { // If the notification is coming from another module
			
			if (notification === "GPIO_ACTION") {
				self.sendSocketNotification(notification, payload);
			}
		}
		
		// Take action on the defined triggers that match the notification parameters
		for (var i = 0; i < self.triggers.length; i++) {
			var trigger = self.triggers[i];
			var notificationMatch, payloadMatch, senderMatch;
			notificationMatch = payloadMatch = senderMatch = false;
			
			if (axis.isUndefined(trigger.notification) || self.deepCompare(trigger.notification, notification)) { notificationMatch = true; }
			if (axis.isUndefined(trigger.sender) || self.deepCompare(trigger.sender, senderName)) { senderMatch = true; }
			if (axis.isUndefined(trigger.payload) || self.deepCompare(trigger.payload, payload)) { payloadMatch = true; }
			
			if (notificationMatch && payloadMatch && senderMatch) {
				for (var k = 0; k < trigger.actions.length; k++) {
					if (trigger.actions[k].action === "NOTIFY") { self.sendNotification(payload.notification, payload.payload); }
					else { self.sendSocketNotification("GPIO_ACTION", trigger.actions[k]); }
				}
			}
		}
	},
	
	/**
	 * Recursively compare two variables to see if they have the same 'value'
	 * Objects are considered qeual if the properties in the first object are also contained in the second. 
	 * 
	 * @param x (any) The first variable to compare
	 * @param y (any) The second variable to compare
	 * @return (boolean) true if the objects are considered 'equal' in 'value', false otherwise
	 */
	deepCompare: function(x, y) {
		var self = this;
		
		// If both references point to the same object or both primitives are equivalent
		if (x === y) { return true; }
		
		// For the following special types, use string representation to compare 'values'
		if	(	( axis.isFunction(x) && axis.isFunction(y) ) ||
				( axis.isDate(x) && axis.isDate(y) ) ||
				( axis.isRegExp(x) && axis.isRegExp(y) ) ||
				( axis.isString(x) && axis.isString(y) ) || // Incase new String() is used
				( axis.isNumber(x) && axis.isNumber(y) ) // Becuase NaN === NaN returns false
			) { return x.toString() === y.toString(); }
		
		if (axis.isObject(x) && axis.isObject(y)) {
			// Deep compare of objects recursively (only comparing if y has the same values as x, not visa versa)
			for (var p in x) {
				if (!x.hasOwnProperty(p)) { continue; }
				if (!y.hasOwnProperty(p)) { return false; }
				return self.deepCompare(x[p], y[p]);
			}
		}
		
		return false;
	},
	
	/**
	 * Override the getDom function to generate the DOM objects to be displayed for this module instance.
	 * Since this module has nothing to display, return an empty text node.
	 */
	getDom: function() { return document.createTextNode(""); },
	
	/**
	 * Override the getScripts function to load additional scripts used by this module. 
	 */
	getScripts: function() {
		var scripts = [];
		if (typeof axis !== "object") { scripts.push(this.file("scripts/axis.js")); }
		return scripts;
	},
	
	/**
	 * Override the getStyles function to load CSS files used by this module. 
	 * Since this module does not display anything, return an empty list. 
	 */
	getStyles: function () { return []; },
	
	/**
	 * Override the getTranslations function to load translation files specific to this module. 
	 */
	getTranslations: function() {
		return {
			en: "translations/en.json",
		};
	},
	
	/**
	 * The log function is a convenience alias that sends a message to the console.  
	 * This is an alias for the MagicMirror Log functions with a developer mode feature added.  
	 * This function prepends the module name to the message.  
	 * 
	 * @param message (string) The message to be sent to the console
	 * @param type (string) The type of message (dev, error, info, log)
	 */
	log: function(message, type) {
		var self = this;
		if (self.config.developerMode) {
			var date = new Date();
			var time = date.getHours() + ":" + date.getMinutes() + ":" + date.getSeconds();
			message = self.name + ": (" + self.data.index + ")(" + time + ") " + message;
		} else { message = self.name + ": " + message; }
		switch (type) {
			case "error": Log.error(message); break;
			case "warn": Log.warn(message); break;
			case "info": Log.info(message); break;
			case "dev": if (self.config.developerMode) { Log.log(message); } break;
			default: Log.log(message);
		}
	}
	
});
